File: solutions/julia/isbn-verifier/isbn-verifier.jl
struct ISBN

    digits::NTuple{10, UInt8}

    function ISBN(s::String)

        if length(s) == 13 && s[2] == s[6] == s[12] == '-'
            s = s[1:1] * s[3:5] * s[7:11] * s[13]
        elseif length(s) != 10
            throw(DomainError(s))
        end

        d = ntuple(i -> begin
            c = s[i]
            if isdigit(c)
                UInt8(c - '0')
            elseif c == 'X' && i == 10
                UInt8(10)
            else
                throw(DomainError("Invalid ISBN character"))
            end
        end, 10)

        is_valid = (d[1] * 10 + d[2] * 9 + d[3] * 8 + d[4] * 7 + d[5] * 6 + d[6] * 5 + d[7] * 4 + d[8] * 3 + d[9] * 2 + d[10] * 1) % 11 == 0

        if !is_valid
            throw(DomainError("Invalid ISBN"))
        end

        return new(d)

    end

end

File: solutions/python/isbn-verifier/isbn_verifier.py
def is_valid(s):

    if len(s) == 13 and (s[1] == s[5] == s[11] == '-'):
        s = s[0] + s[2:5] + s[6:11] + s[12]
    elif len(s) != 10:
        return False

    d = [0 for i in range(0, 10)]

    for i in range(0, 10):
        c = s[i]
        if c.isdigit():
            d[i] = int(c)
        elif i == 9 and c == "X":
            d[i] = 10
        else:
            return False

    return (
        d[0] * 10 + d[1] * 9 + d[2] * 8 + d[3] * 7 + d[4] * 6 +
        d[5] * 5 + d[6] * 4 + d[7] * 3 + d[8] * 2 + d[9] * 1
    ) % 11 == 0
